src/Disciplina/Educator/SizedMerkleTree.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- | Sized Merkle tree implementation.
+module Disciplina.Educator.SizedMerkleTree
+       ( SizedMerkleRoot(..)
+       , SizedMerkleTree (..)
+       , getSizedMerkleRoot
+       , mkSizedMerkleTree
+
+       , SizedMerkleNode (..)
+       , mkBranch
+       , mkLeaf
+       ) where
+
+import Universum hiding (foldMap, toList)
+
+import Disciplina.Crypto (Hash, HasHash, hash, unsafeHash)
+import Disciplina.Crypto.Hash.Class (HashFunc, AbstractHash (..))
+
+import Codec.Serialise (Serialise(..), serialise)
```

Comment:
Why do we have concrete dependencies on hash algorithms and serialisation?

It would be a lot simpler to extract this into separate package, if you abstracted them as interfaces.

Reply:
I'm happy to move this to separate package if @flyingleafe think we should do this.

Reply:
This question was raised several times. We have already agreed to use `Serialise` everywhere as a standard serialisation abstraction.